test/lit/passes/monomorphize-drop.wast

```diff
@@ -0,0 +1,570 @@
+  ;; CAREFUL-NEXT:  (call $work_13)
+  ;; CAREFUL-NEXT: )
+  (func $call-undropped (result i32)
+    ;; As above but now with constant params. We can monomorphize here - there
+    ;; is no issue in optimizing here without a drop and with a drop elsewhere -
+    ;; but we do call a different function of course, that returns an i32.
+    (call $work
+      (i32.const 3)
+      (i32.const 4)
+    )
+  )
+
+  ;; ALWAYS:      (func $call-import (type $0)
+  ;; ALWAYS-NEXT:  (drop
+  ;; ALWAYS-NEXT:   (call $import
+  ;; ALWAYS-NEXT:    (i32.const 3)
+  ;; ALWAYS-NEXT:    (i32.const 4)
+  ;; ALWAYS-NEXT:   )
+  ;; ALWAYS-NEXT:  )
+  ;; ALWAYS-NEXT: )
+  ;; CAREFUL:      (func $call-import (type $0)
+  ;; CAREFUL-NEXT:  (drop
+  ;; CAREFUL-NEXT:   (call $import
+  ;; CAREFUL-NEXT:    (i32.const 3)
+  ;; CAREFUL-NEXT:    (i32.const 4)
+  ;; CAREFUL-NEXT:   )
+  ;; CAREFUL-NEXT:  )
+  ;; CAREFUL-NEXT: )
+  (func $call-import
+    ;; Calling an import allows no optimizations.
+    (drop
+      (call $import
+        (i32.const 3)
+        (i32.const 4)
+      )
+    )
+  )
+
+  ;; ALWAYS:      (func $import-work (type $iii) (param $x i32) (param $y i32) (result i32)
+  ;; ALWAYS-NEXT:  (call $import
+  ;; ALWAYS-NEXT:   (i32.xor
+  ;; ALWAYS-NEXT:    (local.get $x)
+  ;; ALWAYS-NEXT:    (local.get $y)
+  ;; ALWAYS-NEXT:   )
+  ;; ALWAYS-NEXT:   (i32.add
+  ;; ALWAYS-NEXT:    (local.get $x)
+  ;; ALWAYS-NEXT:    (local.get $y)
+  ;; ALWAYS-NEXT:   )
+  ;; ALWAYS-NEXT:  )
+  ;; ALWAYS-NEXT: )
+  ;; CAREFUL:      (func $import-work (type $iii) (param $0 i32) (param $1 i32) (result i32)
+  ;; CAREFUL-NEXT:  (call $import
+  ;; CAREFUL-NEXT:   (i32.xor
+  ;; CAREFUL-NEXT:    (local.get $0)
+  ;; CAREFUL-NEXT:    (local.get $1)
+  ;; CAREFUL-NEXT:   )
+  ;; CAREFUL-NEXT:   (i32.add
+  ;; CAREFUL-NEXT:    (local.get $0)
+  ;; CAREFUL-NEXT:    (local.get $1)
+  ;; CAREFUL-NEXT:   )
+  ;; CAREFUL-NEXT:  )
+  ;; CAREFUL-NEXT: )
+  (func $import-work (param $x i32) (param $y i32) (result i32)
+    ;; Do some work and also call an import.
+    (call $import
+      (i32.xor
+        (local.get $x)
+        (local.get $y)
+      )
+      (i32.add
+        (local.get $x)
+        (local.get $y)
+      )
+    )
+  )
+
+  ;; ALWAYS:      (func $call-import-work (type $0)
+  ;; ALWAYS-NEXT:  (call $import-work_14)
+  ;; ALWAYS-NEXT: )
+  ;; CAREFUL:      (func $call-import-work (type $0)
+  ;; CAREFUL-NEXT:  (drop
+  ;; CAREFUL-NEXT:   (call $import-work
+  ;; CAREFUL-NEXT:    (i32.const 3)
+  ;; CAREFUL-NEXT:    (i32.const 4)
+  ;; CAREFUL-NEXT:   )
+  ;; CAREFUL-NEXT:  )
+  ;; CAREFUL-NEXT: )
+  (func $call-import-work
+    ;; This is monomorphized in ALWAYS (as the drop means the call context is
+    ;; not trivial), but not in CAREFUL mode (as there is no significant benefit
+    ;; from optimizations on the monomorphized function - the import is opaque
+    ;; work we cannot do anything with, even if we see it is handed consts).
```

Comment:
Is precomputing the xor and add not enough of a benefit?

Reply:
Our cost analysis may not be precise enough, but we have this:
```wat
(call $import
 (i32.xor
  (local.get $0)
  (local.get $1)
 )
 (i32.add
  (local.get $0)
  (local.get $1)
 )
)

vs

(drop
 (call $import
  (i32.const 7)
  (i32.const 7)
 )
)
```
The call is the same. `local.get` has 0 cost (we assume it might already be in a register), so that vanishes. And we give consts and super-simple math operations like `i32.add` a cost of 1, so the end up the same.

Reply:
Oh interesting, consts and simple math having the same cost means that we don't prefer replacing math with consts. Could the code size difference tip the scales if it were large enough?

Reply:
We'd need to also measure code size for that, which it looks like we're not doing. In this case, I think that is the right outcome actually: We are duplicating code, so even if we shrink the function we are still growing the total size. So this only makes sense if we actually remove work.

Though in general I think you're right, it's not obvious that considering consts and simple math as the same is right. I think we did simple measurements a decade ago and that seemed ok, but I'm not sure how accurate those were then, much less now... worth revisiting.